presto-orc/src/main/java/com/facebook/presto/orc/writer/LongDictionaryColumnWriter.java

```diff
@@ -0,0 +1,303 @@
+                rowGroupIndexes.set(rowGroupValueCount, UNUSED_VALUE);
+            }
+            else {
+                long value = type.getLong(block, position);
+                int index = dictionary.addIfNotExists(value);
+                rowGroupIndexes.set(rowGroupValueCount, index);
+                statisticsBuilder.addValue(value);
+                rawBytes += typeSize;
+                nonNullValueCount++;
+            }
+            rowGroupValueCount++;
+        }
+        return new BlockStatistics(nonNullValueCount, rawBytes);
+    }
+
+    @Override
+    protected long getRetainedDictionaryBytes()
+    {
+        return INSTANCE_SIZE +
+                dictionary.getRetainedBytes() +
+                dictionaryDataStream.getRetainedBytes() +
+                sizeOf(directValues) +
+                sizeOf(directNulls) +
+                (directColumnWriter == null ? 0 : directColumnWriter.getRetainedBytes());
+    }
+
+    @Override
+    protected Optional<int[]> writeDictionary()
+    {
+        long[] elements = dictionary.elements();
+        for (int i = 0; i < dictionary.size(); i++) {
+            dictionaryDataStream.writeLong(elements[i]);
+        }
+
+        columnEncoding = new ColumnEncoding(DICTIONARY, dictionary.size());
+        return Optional.empty();
+    }
+
+    @Override
+    protected void writePresentAndDataStreams(
+            int rowGroupValueCount,
+            IntBigArray rowGroupIndexes,
+            Optional<int[]> originalDictionaryToSortedIndex,
+            PresentOutputStream presentStream,
+            LongOutputStream dataStream)
+    {
+        checkArgument(!originalDictionaryToSortedIndex.isPresent(), "Unsupported originalDictionaryToSortedIndex");
+        for (int position = 0; position < rowGroupValueCount; position++) {
+            presentStream.writeBoolean(rowGroupIndexes.get(position) != UNUSED_VALUE);
+        }
+        for (int position = 0; position < rowGroupValueCount; position++) {
+            int index = rowGroupIndexes.get(position);
+            // index == UNUSED_VALUE represents null in the dictionary.
+            if (index != UNUSED_VALUE) {
+                dataStream.writeLong(index);
+            }
+        }
+    }
+
+    @Override
+    protected List<StreamDataOutput> getDictionaryStreams(int column)
+    {
+        return ImmutableList.of(dictionaryDataStream.getStreamDataOutput(column));
+    }
+
+    @Override
+    protected ColumnStatistics createColumnStatistics()
+    {
+        ColumnStatistics statistics = statisticsBuilder.buildColumnStatistics();
+        statisticsBuilder = new IntegerStatisticsBuilder();
+        return statistics;
+    }
+
+    @Override
+    protected void closeDictionary()
+    {
+        dictionary = null;
+        dictionaryDataStream.close();
+    }
+
+    @Override
+    protected void resetDictionary()
+    {
+        columnEncoding = null;
+        dictionary = new LongDictionary(10_000);
```

Comment:
Maybe add `clear()` method to `LongDictionary` so we don't have to reallocate space.

Reply:
Some stripes may contain large number of entries in the dictionary and clear would hold on to the same memory. For example SliceDictionaryWriter, re-assigns the element block. Most of the close/reset when there is large number of memory could be allocated, at the end of the stripe, they set to null to start over. So I would prefer to leave it as it is.